src/architecture.c

```diff
@@ -322,37 +322,121 @@ static void arithm(arch_core* core)
+		eff_addr = ((cond_imm_data)core->regs.ir.data).dreg;
+	}
 
+	arch_uint eff_reg = ((cond_imm_data)core->regs.ir.data).breg;
+	arch_word* reg = help_get_reg(core, eff_reg);
+	eff_addr = *help_get_reg(core, eff_addr);
 }
 
 static void add_imm(arch_core* core)
 {
-	
+	arch_uint eff_addr;
+	if (((cond_imm_data)core->regs.ir.data).dreg == 0)
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).addr;
+	}
+	else
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).dreg;
+	}
+
+	arch_uint eff_reg = ((cond_imm_data)core->regs.ir.data).breg;
+	arch_word* reg = help_get_reg(core, eff_reg);
+	eff_addr = *help_get_reg(core, eff_addr);
+	help_get_reg(core, eff_addr).data = (cond_imm_data)core->regs.ir.data).addr + help_get_reg(core, eff_addr).data;
 }
 
 static void mul_imm(arch_core* core)
 {
+	arch_uint eff_addr;
+	if (((cond_imm_data)core->regs.ir.data).dreg == 0)
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).addr;
+	}
+	else
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).dreg;
+	}
 
+	arch_uint eff_reg = ((cond_imm_data)core->regs.ir.data).breg;
+	arch_word* reg = help_get_reg(core, eff_reg);
+	eff_addr = *help_get_reg(core, eff_addr);
+	help_get_reg(core, eff_addr).data = help_get_reg(core, eff_addr).data * (cond_imm_data)core->regs.ir.data).addr;
 }
 
 static void div_imm(arch_core* core)
 {
+	arch_uint eff_addr;
+	if (((cond_imm_data)core->regs.ir.data).dreg == 0)
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).addr;
+	}
+	else
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).dreg;
+	}
 
+	arch_uint eff_reg = ((cond_imm_data)core->regs.ir.data).breg;
+	arch_word* reg = help_get_reg(core, eff_reg);
+	eff_addr = *help_get_reg(core, eff_addr);
+	help_get_reg(core, eff_addr).data = help_get_reg(core, eff_addr).data / (cond_imm_data)core->regs.ir.data).addr;
 }
 
 static void load_imm(arch_core* core)
 {
+	arch_uint eff_addr;
+	if (((cond_imm_data)core->regs.ir.data).dreg == 0)
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).addr;
+	}
+	else
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).dreg;
+	}
 
+	arch_uint eff_reg = ((cond_imm_data)core->regs.ir.data).breg;
+	arch_word* reg = help_get_reg(core, eff_reg);
+	eff_addr = *help_get_reg(core, eff_addr);
+	help_get_reg(core, eff_addr).data = ((cond_imm_data)core->regs.ir.data).addr;
 }
 
 static void set_less_imm(arch_core* core)
 {
+	arch_uint eff_addr;
+	if (((cond_imm_data)core->regs.ir.data).dreg == 0)
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).addr;
+	}
+	else
+	{
+		eff_addr = ((cond_imm_data)core->regs.ir.data).dreg;
+	}
 
+	arch_uint eff_reg = ((cond_imm_data)core->regs.ir.data).breg;
+	arch_word* reg = help_get_reg(core, eff_reg);
+	eff_addr = *help_get_reg(core, eff_addr);
+	if(core->regs.ir.data > ) 
+	{
+		(cond_imm_data)help_get_reg(core, eff_addr).dreg = core=>reg.zr.data + 1;
+	} 
+	else
+	{
+		help_get_reg(core, eff_addr).dreg = core=>reg.zr.data;
+	}
 }
 
 static void halt(arch_core* core)
 {
-
+	core->regs.ac.data = core->(regs.ir.data);
```

Comment:
`halt` will cause an interrupt in our system:
- for all the jobs the professor gave us, `halt` represents the end of the program. For most systems, this means the OS has to reinstate control and exit the program (probably do more things like free space for more programs and start/resume execution of another program).
- because `halt` will do something different depending on the amount of cores running (if it's on the main core, it will cause an interrupt; if it's a thread running on an alternative core, it will freeze the core), its functionality will change.
I (Natan) can go ahead and work on `halt` for you instead.